bpf/process/bpf_rate.h

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Cilium */
+
+#ifndef __RATE_H__
+#define __RATE_H__
+
+#include "bpf_tracing.h"
+#include "bpf_helpers.h"
+
+struct cgroup_rate_key {
+	__u64 id;
+};
+
+struct cgroup_rate_value {
+	__u64 curr;
+	__u64 prev;
+	__u64 time;
+	__u64 rate;
+	__u64 throttled;
+};
+
+struct cgroup_rate_options {
+	__u64 events;
+	__u64 interval;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+	__uint(max_entries, 32768);
+	__type(key, struct cgroup_rate_key);
+	__type(value, struct cgroup_rate_value);
+} cgroup_rate_map SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, __u32);
+	__type(value, struct cgroup_rate_options);
+} cgroup_rate_options_map SEC(".maps");
+
+#define NS (1000ULL * 1000ULL * 1000UL)
+
+static inline __attribute__((always_inline)) bool
+cgroup_rate(void *ctx, struct msg_k8s *kube, __u64 time)
+{
+	struct cgroup_rate_options *opt;
+	struct cgroup_rate_key key = {
+		.id = kube->cgrpid,
+	};
+	struct cgroup_rate_value *val;
+	__u64 delta, interval, slide;
+	__u32 zero = 0;
+
+	opt = map_lookup_elem(&cgroup_rate_options_map, &zero);
+	if (!opt)
+		return true;
+
+	interval = opt->interval;
+	if (!interval)
+		return true;
+
+	val = map_lookup_elem(&cgroup_rate_map, &key);
+	if (!val) {
+		struct cgroup_rate_value new_value = {
+			.time = (time / interval) * interval,
+			.curr = 1,
+		};
+
+		map_update_elem(&cgroup_rate_map, &key, &new_value, 0);
+		return true;
+	}
+
+	/*
+	 * We split the time in interval windows and keep track of events
+	 * of events count in current (val->curr) and previous (val->prev)
+	 * intervals.
+	 */
+
+	delta = time - val->time;
+	if (delta > interval) {
+		if (delta > 2 * interval) {
+			val->prev = 0;
+			val->time = (time / interval) * interval;
+		} else {
+			val->prev = val->curr;
+			val->time += interval;
+		}
+		val->curr = 0;
+	}
+
+	val->curr++;
+
+	/*
+	 * We compute the size of the slide window in previous interval and
+	 * based on that we compute partial amount of events from previous
+	 * interval window. Then we add current interval count and we have
+	 * rate value.
+	 *
+	 *                       val->time
+	 *                       |
+	 *   <--- interval ----->|<--- interval ----->|
+	 *                       |
+	 *    val->prev          | val->curr
+	 *   |-------------------|-----------
+	 *         val->rate
+	 *        |-------------------|
+	 *                            time
+	 */
+
+	slide = interval - (time - val->time);
+	val->rate = (slide * val->prev) / interval + val->curr;
+
+	if (!val->throttled && val->rate >= opt->events)
+		val->throttled = time;
+
```

Comment:
Maybe the throttle start time should be updated every time there is an event received while throttling? So could change the above to:
if (val->rate >= opt->events)
    val->throttled = time;

Also, once val->throttled has a time how does it get reset to 0? Surely it would be easy to check here
if ((time - val->throttled > throttle_duration) && (val->rate < opt->events))
    val->throttled = 0;
Or if you take the suggestion to reset the throttle start time each event (as val->throttled will be updated):
if (time - val->throttled > throttle_duration)
    val->throttled = 0;

Reply:
I see from a later patch that you are summing the rates for the cores, so some of the above can be ignored. That said, if val->throttled is not zero, why not set val->throttled to the current time and reset the throttle duration?

if (val->throttled || val->rate >= opt->events)
val->throttled = time;

Reply:
the logic of the user space code needs to know the time the throttling happened on the cpu, so we record the time just once

Reply:
it's the duty of the user space to monitor the rate once it's throttled, the initial time is there just to make sure we don't get flooded with throttling events in case there are just rate spikes